packages/opentelemetry-core/src/trace/TraceState.ts

```diff
@@ -31,80 +31,83 @@ const LIST_MEMBER_KEY_VALUE_SPLITTER = '=';
-      traceState._internalState.delete(key);
+    if (!validateKey(key) || !validateValue(value)) {
+      // TODO: warn of invalid key or value???
+      return this;
     }
-    traceState._internalState.set(key, value);
-    return traceState;
+
+    const entries = this._getEntries();
+    const index = entries.findIndex(entry => entry.key === key);
+
+    if (index !== -1) {
+      entries.splice(index, 1);
+    }
+    entries.unshift({ key, value });
+
+    const state = this._serializeEntries(entries);
+    return new TraceState(state.length > MAX_TRACE_STATE_LEN ? '' : state);
   }
 
   unset(key: string): TraceState {
-    const traceState = this._clone();
-    traceState._internalState.delete(key);
-    return traceState;
+    const entries = this._getEntries();
+    const index = entries.findIndex(entry => entry.key === key);
+
+    if (index !== -1) {
+      entries.splice(index, 1);
+    }
+
+    return new TraceState(this._serializeEntries(entries));
   }
 
   get(key: string): string | undefined {
-    return this._internalState.get(key);
+    const entries = this._getEntries();
+    const entry = entries.find(e => e.key === key);
+
+    return entry && entry.value;
   }
 
   serialize(): string {
-    return this._keys()
-      .reduce((agg: string[], key) => {
-        agg.push(key + LIST_MEMBER_KEY_VALUE_SPLITTER + this.get(key));
-        return agg;
-      }, [])
-      .join(LIST_MEMBERS_SEPARATOR);
+    return this._state;
   }
 
-  private _parse(rawTraceState: string) {
-    if (rawTraceState.length > MAX_TRACE_STATE_LEN) return;
-    this._internalState = rawTraceState
+  private _getEntries(): Array<{ key: string; value: string }> {
+    // Do not parse/validate if state already surpass max length
+    if (this._state.length > MAX_TRACE_STATE_LEN) {
+      return [];
+    }
+    const entries = [];
+    const members = this._state
       .split(LIST_MEMBERS_SEPARATOR)
-      .reverse() // Store in reverse so new keys (.set(...)) will be placed at the beginning
-      .reduce((agg: Map<string, string>, part: string) => {
-        const listMember = part.trim(); // Optional Whitespace (OWS) handling
-        const i = listMember.indexOf(LIST_MEMBER_KEY_VALUE_SPLITTER);
-        if (i !== -1) {
-          const key = listMember.slice(0, i);
-          const value = listMember.slice(i + 1, part.length);
-          if (validateKey(key) && validateValue(value)) {
-            agg.set(key, value);
-          } else {
-            // TODO: Consider to add warning log
-          }
-        }
-        return agg;
-      }, new Map());
-
-    // Because of the reverse() requirement, trunc must be done after map is created
-    if (this._internalState.size > MAX_TRACE_STATE_ITEMS) {
-      this._internalState = new Map(
-        Array.from(this._internalState.entries())
-          .reverse() // Use reverse same as original tracestate parse chain
-          .slice(0, MAX_TRACE_STATE_ITEMS)
-      );
+      .map(m => m.trim());
```

Comment:
Instead of using `.map(m => m.trim())`, which creates an unnecessary array and adds O(N) complexity, we can trim inside the loop

i.e.

```
for (const m of members) {
        const trimmedMember = m.trim();
```